hudi-client/hudi-client-common/src/test/java/org/apache/hudi/index/bucket/TestPartitionBucketIndexCalculator.java

```diff
@@ -0,0 +1,142 @@
+  }
+
+  /**
+   * Test basic regex rule matching.
+   */
+  @ParameterizedTest(name = "Partition {0} should map to bucket {1}")
+  @MethodSource("providePartitionsAndExpectedBuckets")
+  void testBasicRegexRuleMatching(String partition, int expectedBucket) throws IOException {
+    setUp(DEFAULT_EXPRESSIONS, DEFAULT_RULE, DEFAULT_BUCKET_NUMBER);
+    assertEquals(expectedBucket, calc.computeNumBuckets(partition));
+  }
+
+  /**
+   * Provide test data for parameterized tests.
+   */
+  private static Stream<Arguments> providePartitionsAndExpectedBuckets() {
+    return Stream.of(
+        // Matching the regex pattern - should get bucket 256
+        Arguments.of("2023-06-01", 256),
+        Arguments.of("2023-06-17", 256),
+        Arguments.of("2023-06-18", 256),
+        Arguments.of("2023-11-01", 256),
+        Arguments.of("2023-11-10", 256),
+        Arguments.of("2023-11-11", 256),
+        Arguments.of("2022-06-01", 256),
+        Arguments.of("2021-11-10", 256),
+
+        // Not matching the regex pattern - should get default bucket 10
+        Arguments.of("2023-06-02", 10),
+        Arguments.of("2023-06-19", 10),
+        Arguments.of("2023-07-01", 10),
+        Arguments.of("2023-11-12", 10),
+        Arguments.of("2023-11-02", 10),
+        Arguments.of("2023-05-01", 10),
+        Arguments.of("not-a-date", 10),
+        Arguments.of("", 10)
+    );
+  }
+
+  @Test
+  void testCaching() throws IOException {
+    setUp(DEFAULT_EXPRESSIONS, DEFAULT_RULE, DEFAULT_BUCKET_NUMBER);
+    // Calculate bucket number for the same partition multiple times
+    String partition = "2023-06-01";
+
+    // First call should calculate and cache
+    int bucket1 = calc.computeNumBuckets(partition);
+    assertEquals(256, bucket1);
+    assertEquals(1, calc.getCacheSize());
+
+    // Second call should use the cache
+    int bucket2 = calc.computeNumBuckets(partition);
+    assertEquals(256, bucket2);
+    assertEquals(1, calc.getCacheSize());
+
+    // Different partition should increase cache size
+    calc.computeNumBuckets("2023-11-10");
+    assertEquals(2, calc.getCacheSize());
+  }
+
+  /**
+   * Test multiple regex rules with priority.
+   */
+  @Test
+  void testMultipleRegexRulesWithPriority() throws IOException {
+    String expressions = "\\d{4}-06-01,256;\\d{4}-06-\\d{2},128;\\d{4}-11-\\d{2},64";
+    setUp(expressions, DEFAULT_RULE, DEFAULT_BUCKET_NUMBER);
+    PartitionBucketIndexCalculator calc = PartitionBucketIndexCalculator.getInstance(expressions, DEFAULT_RULE, DEFAULT_BUCKET_NUMBER);
+    // Setup configuration with multiple rules
+
+    // First rule should take priority
+    assertEquals(256, calc.computeNumBuckets("2023-06-01"));
+
+    // Second rule should match when first doesn't
+    assertEquals(128, calc.computeNumBuckets("2023-06-02"));
+    assertEquals(128, calc.computeNumBuckets("2023-06-30"));
+
+    // Third rule should match for November dates
+    assertEquals(64, calc.computeNumBuckets("2023-11-01"));
```

Comment:
Do we test other datatime format like: `2023/11/01` or hive style partition like this: `col1=2023/col2=11/col3=01`

Reply:
all added.